pallets/allocations/src/lib.rs

```diff
@@ -77,84 +82,99 @@ pub mod pallet {
 	pub struct Pallet<T>(PhantomData<T>);
 
 	#[pallet::hooks]
-	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {}
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {
+		fn on_runtime_upgrade() -> Weight {
+			remove_storage_prefix(<Pallet<T>>::name().as_bytes(), b"CoinsConsumed", b"");
+			T::DbWeight::get().writes(1)
+		}
+	}
 
 	#[pallet::call]
 	impl<T: Config> Pallet<T> {
-		/// Can only be called by an oracle, trigger a coin creation and an event
-		#[pallet::weight(
-			<T as pallet::Config>::WeightInfo::allocate(proof.len() as u32)
-		)]
-		// we add the `transactional` modifier here in the event that one of the
-		// transfers fail. the code itself should already prevent this but we add
-		// this as an additional guarantee.
+		/// Optimized allocation call, which will batch allocations of various amounts
+		/// and destinations and together. This allow us to be much more efficient and thus
+		/// increase our chain's capacity in handling these transactions.
+		#[pallet::weight(<T as pallet::Config>::WeightInfo::batch(batch.len().try_into().unwrap_or(T::MaxAllocs::get())))]
 		#[transactional]
-		pub fn allocate(
+		pub fn batch(
 			origin: OriginFor<T>,
-			to: T::AccountId,
-			amount: BalanceOf<T>,
-			proof: Vec<u8>,
+			batch: BoundedVec<(T::AccountId, BalanceOf<T>), T::MaxAllocs>,
 		) -> DispatchResultWithPostInfo {
 			Self::ensure_oracle(origin)?;
-			ensure!(
-				!pallet_emergency_shutdown::Pallet::<T>::shutdown(),
-				Error::<T>::UnderShutdown
-			);
-			ensure!(
-				amount >= T::ExistentialDeposit::get().saturating_mul(2u32.into()),
-				Error::<T>::DoesNotSatisfyExistentialDeposit,
-			);
 
-			if amount == Zero::zero() {
-				return Ok(Pays::No.into());
+			// sanity checks
+			let mut full_issuance: BalanceOf<T> = Zero::zero();
+			for (_account, amount) in batch.iter().cloned() {
+				ensure!(
+					amount >= T::ExistentialDeposit::get().saturating_mul(2u32.into()),
+					Error::<T>::DoesNotSatisfyExistentialDeposit,
+				);
+
+				// overflow, so too many coins to allocate
+				full_issuance = full_issuance
+					.checked_add(&amount)
+					.ok_or(Error::<T>::TooManyCoinsToAllocate)?;
 			}
 
-			let coins_already_allocated = Self::coins_consumed();
-			let coins_that_will_be_consumed = coins_already_allocated
-				.checked_add(&amount)
-				.ok_or("Overflow computing coins consumed")?;
+			if full_issuance == Zero::zero() {
+				return Ok(Pays::No.into());
+			}
 
+			let current_supply = T::Currency::total_issuance();
 			ensure!(
-				coins_that_will_be_consumed <= T::MaximumCoinsEverAllocated::get(),
+				current_supply.saturating_add(full_issuance) <= T::MaximumSupply::get(),
```

Comment:
How about a gradually more aggressive control on the issuance through:
```suggestion
				full_issuance <= T::MaximumSupply::get().saturating_sub(current_supply).div(2u32.into()),
```
I'm just hoping that we can make this check more useful. If you take this suggestion then `use sp_std::ops::Div;` too.

Reply:
Why are we dividing by 2 here?

Reply:
It can be div by any number, but 2 happens to be implemented a lot faster most of the time. The idea is at any single allocation we can only go for a portion of the available supply and not all or close to all.

Reply:
I would rather not do this, or at least not in this PR, since this pallet is supposed to be the only one to issue coins I do not really see the advantage of limiting it in such a way. If we want a cap on allocs (which we probably should), it sounds like something that we'd rather implement in a separate PR, and probably in a very different way (per window limit for instance)

Reply:
Fair enough. We can improve this later with a dedicated PR and I'm also on the side of setting individual window based limits for both the oracle and the destination account.